### ISSUE TITLE
v-for and setTimeout in child component creates index issue

### ISSUE THREAD

**Issue by devboxr** [vuejs/vue#5891]
### Version
2.3.4

### Reproduction link
[https://codepen.io/anon/pen/NgbgRQ?editors=1010](https://codepen.io/anon/pen/NgbgRQ?editors=1010)

### Steps to reproduce
- The CodePen consists of a container and a child component.
- The container owns an array with 3 alerts in it, that are being displayed with `v-for`.
- The child component on mount uses `setTimeout` to schedule an event to let the parent know that it's time to remove it.
- The container listens to the event and removes the item from the array.

When the CodePen opens, the above will execute.

### What is expected?
All 3 alerts should close itself

### What is actually happening?
The alert saying "This should disappear in 5 seconds" continues to stay visible.

---
I did loads of debugging on this one, but the setTimeout() simply doesn't work, while the onClick event works as expected.

**Comment by posva** [comment]
You need to specify a `key` attribute. In that very specific case using `:key="alert.timeout"` works out but you need to find a real id so Vue doesn't reuse the nodes

**Comment by devboxr** [comment]
@posva I understand this and I tried this too, but the index doesn't work (e.g. `:key="index"`).

There is a bigger underlying issue, as the `@click` event uses the right index, while the value inside the `setTimeout()` doesn't. It appears that the scope inside the `setTimeout` (which should be the same as the parent due to arrow function?) isn't notified of the change to the propl

Not sure why this was closed right away :/
